File: DFS/candyDistribution.cpp
#include <iostream>

using namespace std;

int current_count[3] = {2, 1, 1};  // 用來儲存各還有幾顆糖果
int arrangement[4];
string candy[3] = {"紅", "藍", "黃"};

void dfs(int layer){
    if (layer == 3){
        for (int j = 0; j < 4; j++){
            cout << candy[arrangement[j]] << "  ";
        } cout << endl;
    }

    for (int i = 0; i < 3; i++){
        if (current_count[i] == 0){  // 如果沒有剩下了
            continue;
        }
        current_count[i]--;  // 用掉一顆
        arrangement[layer] = i;
        dfs(layer + 1);  // 下一層 DFS
        current_count[i]++;
    }
}

int main(){
    dfs(0);

    return 0;
}

File: DFS/candyDistribution.py
current_count = [2, 1, 1]  # 用來儲存各還有幾顆糖果
arrangement = [''] * 4
candy = {0: '紅', 1: '藍', 2: '黃'}


def dfs(layer):
    if layer == 4:
        print(*arrangement, sep='  ')
        return

    for i in range(3):
        if current_count[i] == 0:  # 如果沒有剩下了
            continue
        current_count[i] -= 1  # 用掉一顆
        arrangement[layer] = candy[i]
        dfs(layer + 1)  # 下一層 DFS
        current_count[i] += 1
        

dfs(0)
